### PR TITLE
[lldb-dap] Refactoring breakpoints to not use the `g_dap` reference.

### PR DESCRIPTION
**Pull request by ashgti**
Refactoring breakpoints to not use the `g_dap` reference.

Instead, when a breakpoint is constructed it will be passed a DAP reference that it should use for its lifetime.

This is part of a larger refactor to remove the global `g_dap` variable to allow us to create multiple DAP instances.

### CODE REVIEW

File: lldb/tools/lldb-dap/BreakpointBase.h
Comment (JDevlieghere) [inline lldb/tools/lldb-dap/BreakpointBase.h:16]: This should follow LLVM's [include style](https://llvm.org/docs/CodingStandards.html#include-style). Applies to all the other files in this PR as well.

Comment (ashgti) [reply lldb/tools/lldb-dap/BreakpointBase.h]: Updated all the files I touched in the CL to match the style.

Comment (ashgti) [reply lldb/tools/lldb-dap/BreakpointBase.h]: I ran `include-what-you-use`[^1] on the breakpoint implementation files and fixed the `#includes` for them. If you think that should go into a separate pull request, I can move that out of here and have a follow up for that as well.

[^1]: https://include-what-you-use.org/

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (labath) [inline lldb/tools/lldb-dap/lldb-dap.cpp:4743]: same comment about container choice

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (labath) [inline lldb/tools/lldb-dap/lldb-dap.cpp:4755]: and braces

File: lldb/tools/lldb-dap/BreakpointBase.h
Comment (labath) [reply lldb/tools/lldb-dap/BreakpointBase.h]: It would have been better off as a separate PR, particularly as this PR contains some not completely trivial refactors in lldb-dap.cpp. I'll just note that the llvm style does not include (for the most part) blank lines between the include groups.

**Review by labath — COMMENTED** [review]
This PR is much easier to review. I mainly focused on the stylistic parts. I'll leave it to the lldb-dap maintainers to review the functionality.

Copying polymorphic objects around is asking for trouble, and I don't think we should be doing that. You didn't come up with that, and it's unrelated to what you're trying to do, so I'm not asking you to change that. I'm just mentioning this in case you feel the urge to do something about.

File: lldb/tools/lldb-dap/InstructionBreakpoint.h
Comment (walter-erquinigo) [reply lldb/tools/lldb-dap/InstructionBreakpoint.h:29]: @labath , you are becoming my reference for high quality code in LLVM.
In general, do you recommend marking most one-argument constructors as explicit in the codebase?

**Review by walter-erquinigo — COMMENTED** [review]
Besides everything mentioned by Pavel, this looks pretty good.

File: lldb/tools/lldb-dap/InstructionBreakpoint.h
Comment (ashgti) [reply lldb/tools/lldb-dap/InstructionBreakpoint.h]: I added `explicit` to the `BreakpointBase::BreakpointBase(DAP &)` constructor and removed all the other unnecessary constructors.

**Comment by ashgti** [comment]
> This PR is much easier to review. I mainly focused on the stylistic parts. I'll leave it to the lldb-dap maintainers to review the functionality.
> 
> Copying polymorphic objects around is asking for trouble, and I don't think we should be doing that. You didn't come up with that, and it's unrelated to what you're trying to do, so I'm not asking you to change that. I'm just mentioning this in case you feel the urge to do something about.

With these changes, and the fact that `BreakpointBase` has a `DAP&` member, the breakpoints cannot be copied. Currently, I construct the instances in place with `try_emplace`/`emplace_back`.

If I did follow up with another refactor on this, do you have any suggestions on how to change that? Should we separate the breakpoint representation from the SBBreakpoint instance?

I do wonder if we should further refactor the types in lldb-dap to be a little more like clangd where we have a more distinct protocol representation with `toJSON`/`fromJSON` JSON serialization support (e.g. https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/Protocol.h). I think that is at least partially the idea behind lldb-dap/JSONUtils.h but that file is more focused on the `toJSON` side of things and the `request_*` methods in lldb-dap.cpp are generally parsing out things as they go (without much in the way of validation or error handling...).

File: lldb/tools/lldb-dap/InstructionBreakpoint.h
Comment (walter-erquinigo) [reply lldb/tools/lldb-dap/InstructionBreakpoint.h]: Thanks for your thoughts!

**Comment by walter-erquinigo** [comment]
> With these changes, and the fact that BreakpointBase has a DAP& member, the breakpoints cannot be copied. 

I don't see that as a problem, but if at some point we need some more flexibility, we should be able to use std::reference_wrapper to store the ref and then allow for copies of objects when it makes sense.

> I do wonder if we should further refactor the types in lldb-dap to be a little more like clangd where we have a more distinct protocol representation with toJSON/fromJSON JSON serialization support (e.g. https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/Protocol.h).

100%!

**Review by walter-erquinigo — APPROVED** [review]
accepting because it seems you addressed all of Pavel's issues

File: lldb/tools/lldb-dap/JSONUtils.cpp
Comment (labath) [inline lldb/tools/lldb-dap/JSONUtils.cpp:1485]: I don't think this change is good as it hides (from the caller) the fact that the function expects the argument to be null terminated. If someone passes a non-null terminated stringref, things will get.. interesting.

You can either keep the C pointer, or do the `.str().c_str()` dance I've seen in some other places.

If, in the first case, you want to get rid of the C strcmp, you can always construct a StringRef locally (`StringRef(key) == "modules"`)

File: lldb/tools/lldb-dap/JSONUtils.h
Comment (labath) [inline lldb/tools/lldb-dap/JSONUtils.h:14]:
```suggestion
#include "lldb/API/SBCompileUnit.h"
```

File: lldb/tools/lldb-dap/SourceBreakpoint.cpp
Comment (labath) [inline lldb/tools/lldb-dap/SourceBreakpoint.cpp:25]:
```suggestion
#include "SourceBreakpoint.h"
#include "BreakpointBase.h"
#include "DAP.h"
#include "JSONUtils.h"
#include "lldb/API/SBBreakpoint.h"
#include "lldb/API/SBFileSpecList.h"
#include "lldb/API/SBFrame.h"
#include "lldb/API/SBTarget.h"
#include "lldb/API/SBThread.h"
#include "lldb/API/SBValue.h"
#include "lldb/lldb-enumerations.h"
#include <cassert>
#include <cctype>
#include <cstdlib>
```

File: lldb/tools/lldb-dap/SourceBreakpoint.h
Comment (labath) [inline lldb/tools/lldb-dap/SourceBreakpoint.h:21]:
```suggestion
#include "DAPForward.h"
#include "lldb/API/SBError.h"
#include "llvm/ADT/StringRef.h"
#include <cstdint>
#include <string>
#include <vector>
```

File: lldb/tools/lldb-dap/Watchpoint.cpp
Comment (labath) [inline lldb/tools/lldb-dap/Watchpoint.cpp:11]:
```suggestion
#include "DAP.h"
```

**Comment by labath** [comment]
> > This PR is much easier to review. I mainly focused on the stylistic parts. I'll leave it to the lldb-dap maintainers to review the functionality.
> > Copying polymorphic objects around is asking for trouble, and I don't think we should be doing that. You didn't come up with that, and it's unrelated to what you're trying to do, so I'm not asking you to change that. I'm just mentioning this in case you feel the urge to do something about.
> 
> With these changes, and the fact that `BreakpointBase` has a `DAP&` member, the breakpoints cannot be copied. Currently, I construct the instances in place with `try_emplace`/`emplace_back`.

Ah, I see. The (unused, I guess) BreakpointBase copy constructor threw me off. In this case, I'm happy.

> 
> If I did follow up with another refactor on this, do you have any suggestions on how to change that? Should we separate the breakpoint representation from the SBBreakpoint instance?
> 
> I do wonder if we should further refactor the types in lldb-dap to be a little more like clangd where we have a more distinct protocol representation with `toJSON`/`fromJSON` JSON serialization support (e.g. https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/Protocol.h). I think that is at least partially the idea behind lldb-dap/JSONUtils.h but that file is more focused on the `toJSON` side of things and the `request_*` methods in lldb-dap.cpp are generally parsing out things as they go (without much in the way of validation or error handling...).

I don't feel I know the code well enough to answer this (at this point, I think you're more familiar with lldb-dap than I am), but I certainly see a lot of room for improvement here. One of the things that really struck me is how all of these breakpoint objects are marked as "structs" -- they have public fields and everything, but then they also have nontrivial internal logic and virtual methods and whatnot. If we want these to be structs, then they should be just data holders and the logic should be somewhere else (I guess that's something along the lines of what you're suggesting), or they should be classes and their data ought to be private.

**Review by labath — APPROVED** [review]

**Comment by walter-erquinigo** [comment]
I remember that when lldb-dap was being brought up, the breakpoint structs were very simple, but they have become non-trivial at this point, so I second the idea of turning them into proper classes with better data encapsulation.

**Pull request merged**
